src/modules/tools/temperaturecontrol/TemperatureControl.cpp

```diff
@@ -529,28 +540,61 @@ void TemperatureControl::on_second_tick(void *argument)
+                    trending_correctly = (get_temperature() > this->runaway_last_temperature);
                 }
-                this->runaway_heating_timer++;
-                if( this->runaway_heating_timer > this->runaway_heating_timeout && this->runaway_heating_timeout != 0 ){
+                else if ( this->runaway_state == COOLING_DOWN ) {
+                    target_reached = (this->get_temperature() <= this->target_temperature);
+                    trending_correctly = (get_temperature() < this->runaway_last_temperature);
+                }
+
+                if ( target_reached ) {
+                    // the temperature has been reached, change to state MAINTAINING_TEMPERATURE
                     this->runaway_heating_timer = 0;
-                    THEKERNEL->streams->printf("ERROR: Temperature took too long to be reached on %s, HALT asserted, TURN POWER OFF IMMEDIATELY - reset or M999 required\n", designator.c_str());
+                    this->runaway_state = MAINTAINING_TEMPERATURE;
+                }
+
+                // as long as the temperature is trending correctly, all is well.
+                // if it stalls, tick the heating timer.
+                if ( trending_correctly ) {
+                    // temperature is trending, reset the timer.
+                    this->runaway_heating_timer = 0;
+                } else {
+                    // temperature is stalled, tick the timer.
+                    this->runaway_heating_timer++;
+                }
+                this->runaway_last_temperature = get_temperature();
+
+                // check whether the temperature has stalled longer than the timeout period.
+                if( this->runaway_heating_timer > this->runaway_heating_timeout && this->runaway_heating_timeout != 0 ) {
+                    this->runaway_heating_timer = 0;
+                    THEKERNEL->streams->printf("ERROR : Temperature took too long to be reached on %s, HALT asserted, TURN POWER OFF IMMEDIATELY - reset or M999 required\n", designator.c_str());
                     THEKERNEL->call_event(ON_HALT, nullptr);
                 }
                 break;
-            case TARGET_TEMPERATURE_REACHED: { // If we are in state TARGET_TEMPERATURE_REACHED, check for thermal runaway
+            }
+            case MAINTAINING_TEMPERATURE: { // handle the case where the target has been reached and temperature is being maintained.
+                // check for thermal runaway
                 float delta= this->get_temperature() - this->target_temperature;
+
                 // If the temperature is outside the acceptable range
-                if(this->runaway_range != 0 && fabsf(delta) > this->runaway_range){
-                    THEKERNEL->streams->printf("ERROR: Temperature runaway on %s (delta temp %f), HALT asserted, TURN POWER OFF IMMEDIATELY - reset or M999 required\n", designator.c_str(), delta);
+                if ( this->runaway_range != 0 && fabsf(delta) > this->runaway_range ) {
```

Comment:
I read the coding guidelines and attempted to set the formatter accordingly.  I'll see about making adjustments to get it right.